File: go/euler31.go
package main
import "fmt"

func main() {
    coins := []int{1, 2, 5, 10, 20, 50, 100, 200}
    dp := [201]int{}
    dp[0] = 1
    
    for _, coin := range coins {
        for i := coin; i < len(dp); i++ {
            dp[i] += dp[i - coin]
        }
    }
    fmt.Println(dp[200])
}

File: ts/euler31.ts
function coinSums(): number {
    const sums: number[] = new Array(201).fill(0);
    sums[0] = 1;
    const coins: number[] = [1, 2, 5, 10, 20, 50, 100, 200];
    for (let coin of coins){
        for (let i = coin; i < sums.length; i++){
            sums[i] += sums[i - coin];
        }
    }

    return sums[200];
}

console.log(coinSums());
